### ISSUE TITLE
Reading Mastering Bitcoin should be a prerequisite, so we can skip the basics

### ISSUE THREAD

**Issue by keblek** [BlockchainCommons/Learning-Bitcoin-from-the-Command-Line#154]
Theres a lot of material in here that while being good would be better delegated to other sources of education. 

Mastering Bitcoin has typically been this resource for many developers and it covers the basics extremely well. 

For example, we shouldn't try to teach someone what an address is, this basic terminology should be delegated to something else.

Therefore, I think this guide would save a lot of editorial effort if it didn't try to teach the basics but rather focused on Linux Systems Administration and how it affects a bitcoin node. 

Focusing on bitcoind and the myriad of ways to administer this service should be what this repo focuses on.

These are just my thoughts and I am looking to start a conversation.

I also want to point out that I don't mean to belittle the efforts of the contributors of this repo but our time is valuable and should be aimed towards adding value instead of repeating things.

**Comment by ChristopherA** [comment]
I agree that LBftCL should have a filter on what prerequisites we ask the readers to have learned, and delegate those learnings elsewhere. 

There was an attempt to at least  briefly review the exact minimum needed, the result which is https://github.com/BlockchainCommons/Learning-Bitcoin-from-the-Command-Line/blob/master/01_0_Introducing_Bitcoin.md — it was written basically 5 years ago after teaching a more basic version of this content at Blockchain University. I think actually most people skip this chapter.

Another filter are things like our course doesn't teach anything about mining — I don't believe that there are many viable jobs or open source projects that require that knowledge. We also are not talking about contributing to bitcoin-core (though we'd love some subset of our graduates of what teach to consider taking the next step to do so.)

The goal is to educate a cadre of people who can USE bitcoin in software development of other layers of the stack, not update bitcoin or create new crypto or adapt for other blockchains. Some evidence to the fact this is working is in this summers batch of interns, most had done some or all of the course, and have come in quite competent in leveraging at least basic RPC to Bitcoin.

This leads me to my problem with @aantonop's  Mastering Bitcoin as a candidate for canonical intro. Having tried to teach from it at Blockchain University it is both too high and too low, and mixes them up a lot. It was a milestone in what it did at the time, but personally I feel is ready for a major refactor and redo of its pedagogy as it tries to do far too much "mastery" when in fact it can't achieve that without doing at least some development. I actually like Bitcoin Clarity by @kiarafbickers for a good separation and balance here, but even then it covers a lot more than what you need to learn to be able to USE bitcoin. I'd love to eventually see something that was more of a chapbook than a text book.

I'm also challenged by some of the python-centric approaches to learning bitcoin. Unfortunately, at the end of these developers often have a sense that they can actually do cryptographic engineering, instead of having the property humility and respect that actually doing it is dangerous and very hard. Talk with @sipa for long enough and you get an appreciation for how hard it is. I don't know how to teach that humility. There are very few even academic programs that teach best practices of secure applied cryptography and adversarial engineering, so examples to learn how to teach this are rare.

I do believe there may be some places in the text past chapter 1 that maybe we are explaining too much. This might be worthy of you pointing out a few specific ones so we can discuss to to make a decision filter to decide if it should be moved to chapter 1 or removed entirely.

I do believe we need to do more about general administration and maintenance of a bitcoin server, including best practices of various settings, however, we have somewhat mitigated that with our bitcoin-standup scripts and tools like Gordian Server. There is a tension on explaining what should be done vs. just offering a script to set it up correctly that we haven't found a great balance for. I think we explained too much before, too soon, but now that we have those scripts a lot of that content about setting up bitcoin is being moved to appendices. @keblek I'd love your thoughts on what we are still missing in the administration category, and where to put it.

In addition to the topic of leveraging prerequisites, we should also identify post-requisites. What should we not teach because it is better taught by a different course that assumes that WE are the pre-requisite.

I appreciate you bringing up the discussion of pedagogy here, and encourage others to pitch in.

-- Christopher Allen

**Comment by keblek** [comment]
Hi Christopher, 

I tried my best to summarise my thoughts, please understand that I dont have any experience teaching Bitcoin or computer security. In fact I am still a CS student, so these are just my thoughts.

TLDR: Make the coursetowards finishing tasks, tell them what they need to do not how to do it. Then show them how you would do it. The course should be a hollistic approach on how to bootstrap a robust bitcoin based service; which necessitates a bitcoin node and all the baggage that comes with it. 

> our course doesn't teach anything about mining

I don't think it should especially since you'd need physical access to a lab to really learn mining since its mostly a matter of Mechanical Engineering (thermo, power management), there is also the operations (how to run a factory that inputs electricity and computers and outputs blocks) and Risk Management side which is itself another topic. Unless we are talking about mining pool protocols or distributed consensus algorithms in general. Mining is its own rabbit hole and deserves its own repo. Miners should be running their own nodes, so maybe this can be helpful in this way.

> We also are not talking about contributing to bitcoin-core

I think a separate repo on how to contribute to Bitcoin Core would be a great addition! A good place to start would be how to help with reviewing and how to run (and write) tests. 

The Bitcoin Clarity website looks like a good resource thanks for sharing.

> I'm also challenged by some of the python-centric approaches to learning bitcoin. 

You must be referring to Programming Bitcoin by @jimmysong? I am actually halfway through and am really enjoying it - although you'd have to be naive to think that you can call yourself a cryptographer after having finished it. If someone is this naive about how nuanced something like cryptography is then its unlikely that they'll get their hands on other peoples money or write-access to the Bitcoin Core repo. I like to think that society at large has filters for these kinds of people, so I don't see it as a problem. Thankfully, Bitcoin (network and community) will humble these naive individuals if they don't humble themselves.  

This repo briefly covers the math behind Bitcoin well enough to get the ball rolling - so I think we can link to good vetted sources of information that explain it a bit further. Our goal should not be to make cryptographers but for the graduates to at least be able to hold a conversation with an experienced core or lightning developer (so that they can help them - rather than take their time). Not everything about Bitcoin is about cryptography, the P2P engine is a critical component of the client which doesn't rely on signatures or hashes.

> I do believe there may be some places in the text past chapter 1 that maybe we are explaining too much. 

Yes, take a look at 3.3 "Setting up your wallet". I firmly believe that if you're doing this course and you have never heard of what an address is then you should go read the first few chapters of Mastering Bitcoin (or watch a youtube video or something). I seriously doubt someone encountering this terminology for the first time can clearly articulate why it is that they are spending their time wisely learning about Bitcoin. 

Having said that, it is important to explain how bitcoind manages wallets and where addresses are stored and how to interact with them productively. In short, we should assume that they have at least made a bitcoin transaction once in their life - if not they should go download electrum and get either some real bitcoin or some testnet bitcoin. There is no good way to teach this through a markdown document - they must do these first few leaps on their own.

I am making my way through this lesson plan and will make small PR's to trim out some of the text, I strongly believe that it will make this repo more maintainable.

> I do believe we need to do more about general administration and maintenance of a bitcoin server

I strongly agree. Even to a CS bachelor this task can be daunting as this is where the rubber meets the road (especially with regards to setting up a robust service). I would like to think that someone that has finished this course can productively contribute to myNode, raspiblitz, Gordian (StandUp), bitcoin-fort and bitcoin nodes setups in general so that access to a secure Bitcoin node is made as accessible as possible (we really need as much of this as possible). 

We could even try to collaborate with raspiblitz and have some sort of program that helps you to understand bitcoind, similar to [rustlings](https://github.com/rust-lang/rustlings). It's a rust program that watches files on an infinite loop, something like this but for a bitcoin node might be a welcome resource for the community. 

They can also use this experience to assist their local bitcoin exchange and help with running a node - many exchanges simply don't consider running a node a core competency and thats probably because there isn't enough talent out there to help with this. So they outsource it to blockchain.info - this aspect of the space really bad IMO. I blame shitcoins for this, since it is very complicated for an exchange to concurrently manage multiple blockchains.

I think an annotated version of the unix standup script would be welcome and to have the student write each command one-by-one and explain why its necessary, would be a good lesson. For example, explaining why do we need to create a new user for bitcoin? Why are we checking signatures? Should I install through the package manager or compile from source? and what is the difference? 

Having said that, I think the SysAdmin sections that we do already have (as they relate to bitcoin) should be descriptive, not prescriptive. 

For example, there is a section on jq (a fine tool and a fine section) but the students should just be told that they need to query bitcoind and process the reply and then we show them how we would do it after they thought about it. Give them some links to some tools and let them figure it out. I think its great that we are learning shell since its ubiquitous but maybe someone would rather make http requests and process the reply using the python standard library (or Go). I personally dislike working with sets and lists in shell, so I tend to avoid it.

I should also point out that I am struggling to think of where to draw the line with how much detail we include. 

For example, when you curl a block explorer to get transaction info you have to know to redirect the response in order to work with the data (if you just wget the URL you won't see text in your terminal). So should we explain that? What a shell is, how unix works, everything being a file? Where do we draw the line? So if someone could step in and provide some guidance on this I would appreciate it.

With regards to SysAdmin work, I think we should emphasise that they learn how to understand how the networking stack works on their machines - on their own. Given an linux instance thats running bitcoind (for example - could also be a web server) how would you query how many system resources it's consuming? What does using those resources mean? Could it ever be a problem? How would you fix it? These are things I am in the process of learning (things like ps, nmap, htop) as well - I am learning this because I personally don't want to just copy and paste commands into a shell and cross my fingers that what I think is happening is actually happening. 

Another good topic to teach in here that we can touch on personally is what a home network looks like (in the average case). This is where I think we can have the most impact and add the most value. Both because it is a very tricky topic (different home network configurations) but also because many bitcoin nodes (I would hope) are running in someones home. Subtopics might include IPv4, ports, firewalls, what is a public IP, ASN (if we are feeling ambitious) and TOR/Hidden Services.

A good lesson plan would be to teach how to connect c-lightning in the cloud to a bitcoin node running at home, or how to have a bitcoin and lightning (configs, channels, blocks) backup system running at home over a network, or how to make a watcher program that bootstraps a node (with carryover) in the cloud if your local node disconnects. This would certainly be a challenge and I recognise that it is borderline overly-ambitious.

Another idea for a topic that we can include is File Systems or at least how a FS relates to the blockchain and channel management - I know that they don't matter that much since the blockchain is durable enough and even with lightning it (a perfect file system) doesn't help because you have to figure out the state of the channel when you reboot from a crash anyways. 

Having said that - I think its valuable to know how to make a good backup of the blockchain (since IBD is so very tedious) and how to manage it. What happens if I run out of disk space (how can I increase it with minimal downtime)? How do I separate my .bitcoin directory so that bitcoind can take advantage of my small (but fast) SSD while my blockchain can live in large (but slow) HDD? What tools do I have to make this work (EXT4, BTRFS, ZFS, APFS, NTFS, IPFS or In-Memory FS's, running on SD cards, SSD's, HDD's)?. 

I think this material can go in the beginning - before we show them how to install bitcoind -  so that those that are familiar with it can skip it and those that are newcomers can start experimenting right away since lots of this stuff is bitcoin agnostic. 

I mentioned package management previously, I think it should be briefly covered, since there are multiple ways to download packages. Each with its pros and cons.

Having said all of this, I recognise that it represents a significant amount of work that is nuanced and diverse. I also have not laid out how to make any of my suggestions a reality. So I just want to make it clear that I want to learn as much as I can and I want to help other learn what I learned. I can contribute but I am by no means an expert on any of these topics (I am still enrolled in a CS program), but if I can approximate a level of expert knowledge or help others then I will gladly do so.

I look forward to hearing anyones feedback and I hope this wall of text is not too tall.

**Comment by ChristopherA** [comment]
Thanks your for your pedagogy advice. We are taking it into consideration as we move forward our work on the course.

-- Christopher Allen
